src/dask_awkward/lib/optimize.py

```diff
@@ -419,23 +420,31 @@ def _get_column_reports(dsk: HighLevelGraph) -> dict[str, Any]:
 
 def _necessary_columns(dsk: HighLevelGraph) -> dict[str, list[str]]:
     """Pair layer names with lists of necessary columns."""
-    kv = {}
+    layer_to_columns = {}
     for name, report in _get_column_reports(dsk).items():
-        cols = {_ for _ in report.data_touched if _ is not None}
-        select = []
-        for col in sorted(cols):
-            if col == name:
+        touched_data_keys = {_ for _ in report.data_touched if _ is not None}
+
+        necessary_columns = []
+        for key in sorted(touched_data_keys):
+            if key == name:
+                continue
+
+            layer, column = key.split(".", 1)
+            if layer != name:
                 continue
-            n, c = col.split(".", 1)
-            if n == name:
-                if c.endswith("__list__"):
-                    cnew = c[:-9].rstrip(".")
-                    if cnew not in select:
-                        select.append(f"{cnew}.*")
-                else:
-                    select.append(c)
-        kv[name] = select
-    return kv
+
+            # List offsets are tagged as {key}.{LIST_KEY}. This routine resolve
+            # _columns_, so we use a wildcard to indicate that we want to load
+            # *any* child column of this list. If the list contains no records,
+            # then we load
+            if column.endswith(LIST_KEY):
+                list_parent_path = column[: -(len(LIST_KEY) + 1)].rstrip(".")
+                if list_parent_path not in necessary_columns:
+                    necessary_columns.append(f"{list_parent_path}.*")
+            else:
+                necessary_columns.append(column)
```

Comment:
@douglasdavis this is where I was initially thinking that we need to do something for `shape_touched`. However, it might prove fiddly.

Naively, I thought about `A.*` if we want the shape of `A.B`. However, this wouldn't work for the following type of `A`
```
1 * {
    x: var * int64,
    y: var * int64
}
```

Computing `A.y` wouldn't give any information about the length of the `NumpyArray` in `A.x`.

Reply:
Just to follow up (I don't expect you to have seen this over the weekend) — this is the blocker for @lgray's demo this week AFAICT.

Reply:
This PR fixes the pickling problem, but there's still the task of loading lengths for buffers with `shape_touched`. I'll move this to Slack :)